src/plugins/BaseObjectSerializer.ts

```diff
@@ -0,0 +1,138 @@
+import { MD5, enc } from 'crypto-js'
+
+/**
+ * Serializer for Speckle objects written in Typescript
+ */
+export class BaseObjectSerializer {
+  constructor(public transports: ITransport[]) {}
+
+  public async SerializeBase(object: object): Promise<SerializedBase> {
+    return await this.SerializeBaseWithClosures(object, [])
+  }
+
+  private async SerializeBaseWithClosures(object: object, closures: Array<Map<string, number>>) {
+    const thisClosure = new Map<string, number>()
+    closures.push(thisClosure)
+
+    const converted = await this.PreserializeEachObjectProperty(object, closures)
+    let json = this.SerializeMap(converted)
+    const id = this.GetId(json)
+    converted.set('id', id)
+
+    this.AddSelfToParentClosures(id, closures)
+    if (thisClosure.size > 0) {
+      converted.set('__closure', Object.fromEntries(thisClosure))
+    }
+    converted.set('totalChildrenCount', thisClosure.size)
+
+    json = this.SerializeMap(converted)
+    await this.StoreObject(id, converted)
+    return new SerializedBase(id, json)
+  }
+
+  private async PreserializeObject(
+    object: any,
+    closures: Array<Map<string, number>>
+  ): Promise<any> {
+    if (!(object instanceof Object) || object instanceof String) {
+      return object
+    }
+
+    if (object instanceof DataChunk) {
+      const serialized = await this.SerializeBaseWithClosures(object, [...closures])
+      return new ObjectReference(serialized.id)
+    }
+
+    if (object instanceof Array) {
+      // chunk array into 5000 by default
+      const chunkSize = 5000
+      if (object.length > chunkSize) {
+        let serializedCount = 0
+        const data = new Array<DataChunk>()
+        while (serializedCount < object.length) {
+          const dataChunkCount = Math.min(chunkSize, object.length - serializedCount)
+          data.push(new DataChunk(object.slice(serializedCount, serializedCount + dataChunkCount)))
+          serializedCount += dataChunkCount
+        }
+        return await this.PreserializeObject(data, closures)
+      }
+
+      const convertedList = new Array<any>()
+      for (const element of object) {
+        convertedList.push(await this.PreserializeObject(element, closures))
+      }
+      return convertedList
+    }
+
+    if (object instanceof Object) {
+      return Object.fromEntries(await this.PreserializeEachObjectProperty(object, closures))
+    }
+
+    throw new Error(`Cannot serialize object ${object}`)
+  }
+
+  private async PreserializeEachObjectProperty(
+    o: object,
+    closures: Array<Map<string, number>>
+  ): Promise<Map<string, any>> {
+    const converted = new Map<string, any>()
+
+    for (const key of Object.keys(o)) {
+      const objKey = key as keyof object
+      converted.set(key, await this.PreserializeObject(o[objKey], closures))
```

Comment:
A key part here I notice missing is the lack of `key` validation, as we do have some non-allowed characters in Speckle object properties.